### PR TITLE
[Feature] Shadows

### PR DESCRIPTION
**Pull request by chMoussa**
- [x] Porting shadow from Qadence
- [x] Harmonize tomography and shadow `compute_expectation`
- [x] Add tests
- [x] Modify docs

### CODE REVIEW

File: docs/measurements/index.md [hunk 1]
```diff
@@ -61,7 +61,34 @@ print(f"Estimated expectation value tomo = {estimated_values_tomo}") # markdown-
 
 ## Classical shadows
 
-Coming soon!
+A much less resource demanding protocol based on _classical shadows_ has been proposed[^1]. It combines ideas from shadow tomography[^2] and randomized measurement protocols capable of learning a classical shadow of an unknown quantum state $\rho$. It relies on deliberately discarding the full classical characterization of the quantum state, and instead focuses on accurately predicting a restricted set of properties that provide efficient protocols for the study of the system.
```
Comment (RolandMacDoland) [inline docs/measurements/index.md]:
```suggestion
A much less resource demanding protocol based on _classical shadows_ has been proposed[^1]. It combines ideas from shadow tomography[^2] and randomized measurement protocols [ref.?] capable of learning a classical shadow of an unknown quantum state $\rho$. It relies on deliberately discarding the full classical characterization of the quantum state, and instead focuses on accurately predicting a restricted set of properties that provide efficient resources for the study of the system.
```

File: qadence_protocols/measurements/protocols.py [hunk 2]
```diff
@@ -39,6 +42,27 @@ def __call__(
             module = importlib.import_module(PROTOCOL_TO_MODULE[self.protocol])
         except (KeyError, ModuleNotFoundError, ImportError) as e:
             raise type(e)(f"Failed to import Mitigations due to {e}.")
-        # Partially pass the options.
-        expectation = partial(getattr(module, "compute_expectation"), options=self.options)
+
+        conv_observables = model._observable
+        if not isinstance(conv_observables, list):
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: Doesn't the `._observable` property return a list ?

File: qadence_protocols/measurements/protocols.py [hunk 3]
```diff
@@ -39,6 +42,27 @@ def __call__(
...
+            raise TypeError(
+                "Observables must be of type <class 'list[ConvertedObservable]'>. Got {}.".format(
+                    type(conv_observables)
+                )
+            )
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]:
```suggestion
                f"Observables must be of type <class 'list[ConvertedObservable]'>. Got { type(conv_observables)}."
            )
```

File: qadence_protocols/measurements/protocols.py [hunk 4]
```diff
@@ -39,6 +42,27 @@ def __call__(
...
+
+        try:
+            observables = [obs.abstract for obs in conv_observables]
+        except AttributeError:
+            raise TypeError(
+                "Observables must be of type <class 'list[ConvertedObservable]'>. Got {}.".format(
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/protocols.py]: Same.

File: qadence_protocols/measurements/shadow.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    shadow_size = options.get("shadow_size", None)
+    accuracy = options.get("accuracy", None)
+    if shadow_size is None and accuracy is None:
+        KeyError(
+            "Shadow protocol requires either an option"
+            "'shadow_size' of type 'int' or 'accuracy' of type 'float'."
+        )
+    confidence = options.get("confidence", None)
+    if confidence is None:
+        KeyError("Shadow protocol requires a 'confidence' kwarg of type 'float'.")
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/shadow.py]:
```suggestion
        KeyError("Shadow protocol requires an option 'confidence' of type 'float'.")
```

File: qadence_protocols/measurements/utils_shadow.py [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    return max([_max_observable_weight(observable=observable) for observable in observables])
+
+
+def number_of_samples(
+    observables: list[AbstractBlock], accuracy: float, confidence: float
+) -> tuple[int, ...]:
+    """
+    Estimate an optimal shot budget and a shadow partition size.
+
+    to guarantee given accuracy on all observables expectation values
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow.py]:
```suggestion
    This is to guarantee given accuracy on all observables expectation values
```

File: qadence_protocols/measurements/utils_shadow.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    N: int,
+    K: int,
+    shadow: list,
+    observable: AbstractBlock,
+    endianness: Endianness = Endianness.BIG,
+) -> Tensor:
+    """
+    Return estimators (traces of observable times mean density matrix).
+
+    for K equally-sized shadow partitions.
```
Comment (RolandMacDoland) [inline qadence_protocols/measurements/utils_shadow.py]:
```suggestion
    Estimators are computed for K equally-sized shadow partitions.
```

File: tests/test_shadow.py [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+        qubit_support=circuit.block.qubit_support,
+        N=N,
+        K=K,
+        shadow=shadows[0],
+        observable=observable,
+    )
+    assert torch.allclose(estimated_traces, exp_traces)
+
+
+# @pytest.mark.flaky(max_runs=5)
```
Comment (RolandMacDoland) [inline tests/test_shadow.py]: Isn't it possible to enable these tests ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. Looking good. Some comments from my side.

File: qadence_protocols/measurements/protocols.py (hunk 2 above)
Comment (chMoussa) [reply qadence_protocols/measurements/protocols.py]: Indeed, I'll remove this part.

File: tests/test_shadow.py (hunk 8 above)
Comment (chMoussa) [reply tests/test_shadow.py]: Yes I forgot to uncomment when testing as these are long.

File: qadence_protocols/measurements/utils_shadow.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+        return reduce(torch.kron, local_density_matrices)
+
+
+def classical_shadow(
+    shadow_size: int,
+    circuit: QuantumCircuit,
+    param_values: dict,
+    state: Tensor | None = None,
+    backend: Backend | DifferentiableBackend = PyQBackend(),
+    # FIXME: Changed below from Little to Big, double-check when Roland is back
```
Comment (jpmoutinho) [inline qadence_protocols/measurements/utils_shadow.py]: Well I think you can remove this comment which has been there for many months, I assume the results are correct now 😅

File: qadence_protocols/measurements/utils_shadow.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+    observable: AbstractBlock,
+    endianness: Endianness = Endianness.BIG,
+) -> list:
+    floor = int(np.floor(N / K))
+    traces = []
+    # TODO: Parallelize embarrassingly parallel loop.
+    for k in range(K):
+        reconstructed_state = reconstruct_state(shadow=shadow[k * floor : (k + 1) * floor])
+        # Reshape the observable matrix to fit the density matrix dimensions
+        # by filling indentites.
```
Comment (jpmoutinho) [inline qadence_protocols/measurements/utils_shadow.py:176]: This is probably one of the major performance bottlenecks, maybe add an issue that this needs to be solved without explicit identities? It seems similar to what is being done in PyQ for the noise.

**Review by jpmoutinho — APPROVED** [review]
Thanks @chMoussa. Not much to say since I am not very familiar with this code. Since you looked at it in more detail, the main thing I would suggest is to open some issues in case you spotted parts that can be optimized.

File: qadence_protocols/measurements/utils_shadow.py (hunk 10 above)
Comment (chMoussa) [reply qadence_protocols/measurements/utils_shadow.py:176]: Ok opened issue #39

**Pull request merged**